Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/tangential_relaxation.h

```diff
@@ -329,6 +329,304 @@ void tangential_relaxation(const VertexRange& vertices,
+
+  typedef typename GT::Vector_3 Vector_3;
+  typedef typename GT::Point_3 Point_3;
+
+  auto check_normals = [&](vertex_descriptor v)
+  {
+    bool first_run = true;
+    Vector_3 prev = NULL_VECTOR, first = NULL_VECTOR;
+    halfedge_descriptor first_h = boost::graph_traits<TriangleMesh>::null_halfedge();
+    for (halfedge_descriptor hd : CGAL::halfedges_around_target(v, tm))
+    {
+      if (is_border(hd, tm)) continue;
+
+      Vector_3 n = compute_face_normal(face(hd, tm), tm, np);
+      if (n == CGAL::NULL_VECTOR) //for degenerate faces
+        continue;
+
+      if (first_run)
+      {
+        first_run = false;
+        first = n;
+        first_h = hd;
+      }
+      else
+      {
+        if (!get(ecm, edge(hd, tm)))
+          if (to_double(n * prev) <= 0)
+            return false;
+      }
+      prev = n;
+    }
+
+    if (first_run)
+      return true; //vertex incident only to degenerate faces
+
+    if (!get(ecm, edge(first_h, tm)))
+      if (to_double(first * prev) <= 0)
+        return false;
+
+    return true;
+  };
+
+  typedef typename internal_np::Lookup_named_param_def <
+    internal_np::allow_move_functor_t,
+    NamedParameters,
+    internal::Allow_all_moves// default
+  > ::type Shall_move;
+  Shall_move shall_move = choose_parameter(get_parameter(np, internal_np::allow_move_functor),
+    internal::Allow_all_moves());
+
+  for (unsigned int nit = 0; nit < nb_iterations; ++nit)
+  {
+#ifdef CGAL_PMP_TANGENTIAL_RELAXATION_VERBOSE
+    std::cout << "\r\t(Tangential relaxation iteration " << (nit + 1) << " / ";
+    std::cout << nb_iterations << ") ";
+    std::cout.flush();
+#endif
+
+    typedef std::tuple<vertex_descriptor, Vector_3, Point_3> VNP;
+    std::vector< VNP > barycenters;
+    auto gt_barycenter = gt.construct_barycenter_3_object();
+    auto gt_centroid = gt.construct_centroid_3_object();
+    auto gt_area = gt.compute_area_3_object();
+
+    // at each vertex, compute vertex normal
+    std::unordered_map<vertex_descriptor, Vector_3> vnormals;
+    compute_vertex_normals(tm, boost::make_assoc_property_map(vnormals), np);
+
+    // at each vertex, compute centroids of neighbouring faces weighted by
+    // area and sizing field
+    for(vertex_descriptor v : vertices)
+    {
+      if (get(vcm, v) || CGAL::internal::is_isolated(v, tm))
+        continue;
+
+      // collect hedges to detect if we have to handle boundary cases
+      std::vector<halfedge_descriptor> interior_hedges, border_halfedges;
+      for(halfedge_descriptor h : halfedges_around_target(v, tm))
+      {
+        if (is_border_edge(h, tm) || get(ecm, edge(h, tm)))
+          border_halfedges.push_back(h);
+        else
+          interior_hedges.push_back(h);
+      }
+
+      if (border_halfedges.empty())
+      {
+        const Vector_3& vn = vnormals.at(v);
+        Vector_3 move = CGAL::NULL_VECTOR;
+        double weight = 0;
+        for(halfedge_descriptor h :interior_hedges)
+        {
+          // calculate weight
```

Comment:
Should that be added, or does it already exist in the  [Weights](https://doc.cgal.org/latest/Weights/group__PkgWeightsRef.html) package?